### PR TITLE
Fix High impact vulnerability on remote executors

### PR DESCRIPTION
**Pull request by nnfrog**
This PR is meant to address the security issue we've also reported over email.
The change attempts to replace all pickle usages with JSON serialization, in order to avoid unwanted code execution.
Would be great to get another set of eyes on it, as we're not sure we've tested all possible scenarios.
The usage of the Pickle library is extremely dangerous and it is recommended to consider any alternative before using it.

### CODE REVIEW

**Comment by nnfrog** [comment]
@albertvillanova I answered to the Copilot review. How do we advance the process from here - do I ask you to activate a re-review from the Copilot or it happens automatically?

File: src/smolagents/remote_executors.py [hunk 1]
```diff
@@ -65,36 +64,131 @@ def run_code_raise_errors(self, code: str) -> CodeOutput:
...
+        encoded_vars = base64.b64encode(json.dumps(json_safe_vars).encode()).decode()
+        
         code = f"""
-import pickle, base64
-vars_dict = pickle.loads(base64.b64decode('{pickled_vars}'))
+import json, base64
+import numpy as np
+from datetime import datetime
+from io import BytesIO
+import PIL.Image
```
[the remark this answers is not in the document]
Comment (albertvillanova) [reply src/smolagents/remote_executors.py]: Forget Copilot comment.

You are assuming that the executor has `numpy` and `pillow` installed, but this might not be the case.

File: src/smolagents/remote_executors.py [hunk 2]
```diff
@@ -65,36 +64,131 @@ def run_code_raise_errors(self, code: str) -> CodeOutput:
...
+            return {"__type__": "tuple", "data": [self._to_json_safe(item) for item in obj]}
+        elif isinstance(obj, set):
+            return {"__type__": "set", "data": [self._to_json_safe(item) for item in obj]}
+        elif isinstance(obj, bytes):
+            return {"__type__": "bytes", "data": base64.b64encode(obj).decode()}
+        elif hasattr(obj, 'tolist'):  # numpy arrays
+            return {"__type__": "ndarray", "data": obj.tolist(), "dtype": str(obj.dtype)}
+        elif hasattr(obj, 'isoformat'):  # datetime objects
+            return {"__type__": "datetime", "data": obj.isoformat()}
+        elif hasattr(obj, 'save'):  # PIL Images
```
Comment (albertvillanova) [inline src/smolagents/remote_executors.py]: Not sure about these last 3 conditions: other classes may implement these methods, which could lead to false positives and wrong serialization. I think it would be safer and more explicit to use `isinstance` checks against the expected classes (with optional imports guarded).

File: src/smolagents/remote_executors.py [hunk 3]
```diff
@@ -65,36 +64,131 @@ def run_code_raise_errors(self, code: str) -> CodeOutput:
...
+            raise TypeError(f"Object of type {type(obj).__name__} is not JSON serializable")
+    
+    def _from_json_safe(self, obj):
+        """Convert JSON-safe format back to Python objects."""
+        if isinstance(obj, dict):
+            if "__type__" in obj:
+                obj_type = obj["__type__"]
+                if obj_type == "ndarray":
+                    try:
+                        import numpy as np
```
Comment (albertvillanova) [inline src/smolagents/remote_executors.py]: Here you are assuming `numpy` is installed locally, but this might not be the case.

File: src/smolagents/remote_executors.py [hunk 4]
```diff
@@ -65,36 +64,131 @@ def run_code_raise_errors(self, code: str) -> CodeOutput:
...
-        if "final_answer" in tools:
-            self._patch_final_answer_with_exception(tools["final_answer"])
+        # Filter out non-Tool objects (like executor instances from managed_agents)
+        try:
+            from smolagents.tools import Tool as ToolClass
+        except ImportError:
+            # Fallback for different import paths
+            from .tools import Tool as ToolClass
+        
+        filtered_tools = {name: tool for name, tool in tools.items() if isinstance(tool, ToolClass)}
```
Comment (albertvillanova) [inline src/smolagents/remote_executors.py]: Why did you add this filter? Not sure of fully understanding it...

File: src/smolagents/remote_executors.py [hunk 5]
```diff
@@ -127,13 +221,44 @@ class _FinalAnswerTool(final_answer_tool.__class__):
...
+                    return {"__type__": "ndarray", "data": obj.tolist(), "dtype": str(obj.dtype)}
+                elif hasattr(obj, 'isoformat'):  # datetime objects
+                    return {"__type__": "datetime", "data": obj.isoformat()}
+                elif hasattr(obj, 'save'):  # PIL Images
+                    buffer = BytesIO()
+                    obj.save(buffer, format='PNG')
+                    return {"__type__": "PIL.Image", "data": base64.b64encode(buffer.getvalue()).decode()}
+                else:
+                    # For unsupported types, convert to string representation
+                    return str(obj)
```
Comment (albertvillanova) [inline src/smolagents/remote_executors.py]: You already defined `_to_json_safe` above and here again.

**Review by albertvillanova — CHANGES_REQUESTED** [review]
Thanks for tackling the pickle issue: removing unsafe deserialization from remote executors is a right move.

Note that originally, the remote executor code assumed a trusted threat model where the agent creator was considered benign: the creator defines which tools can be invoked and which model drives the agent, so it was expected that no malicious payloads would ever flow through execution. In practice, however, this assumption is too weak: once execution crosses process or machine boundaries, data could be compromised. Under those conditions, using pickle meant deserializing arbitrary byte streams with full code execution privileges, a well-known high-impact vulnerability. 

This PR improves this model by replacing pickle with a controlled, JSON-based serialization that handles only specific types (arrays, bytes, images, etc.), eliminating the arbitrary code execution risk and making remote execution safe by default.

Just some comments below (see also inline comments):
- Duplication of serialization logic in multiple places:
  - `RemotePythonExecutor._to_json_safe()`
  - `RemotePythonExecutor._from_json_safe()`
  - Embedded in `_patch_final_answer_with_exception()`
  - String-generated code in `send_variables()
- Missing import guards: code assumes some libs are always available

Additionally, we could improve the test coverage. Some ideas:
- Round-trip type fidelity validation
- Error handling with missing dependencies
- Some edge cases?

**Comment by nnfrog** [comment]
@albertvillanova I made a new commit with a more appealing implementation
I think I addresses most of the issues you mentioned.
Now all the des/ser code is on centralized class,
I removed uneeded imports
check the PIL with isinstance

WDYT?

**Comment by nnfrog** [comment]
I resolved the conflicts

File: src/smolagents/remote_executors.py [hunk 6]
```diff
@@ -249,6 +340,93 @@ def cleanup(self):
         except Exception as e:
             self.logger.log_error(f"Error during cleanup: {e}")
 
+def _websocket_send_execute_request(code: str, ws) -> str:
```
Comment (albertvillanova) [inline src/smolagents/remote_executors.py]: I think you introduced duplicate functions while resolving the conflicts with the main branch: this function was already defined and it is present in line 431.

File: src/smolagents/remote_executors.py [hunk 7]
```diff
@@ -249,6 +340,93 @@ def cleanup(self):
...
+            "store_history": True,
+            "user_expressions": {},
+            "allow_stdin": False,
+        },
+    }
+
+    ws.send(json.dumps(execute_request))
+    return msg_id
+
+def _websocket_run_code_raise_errors(code: str, ws, logger) -> CodeOutput:
```
Comment (albertvillanova) [inline src/smolagents/remote_executors.py]: The same for this function.

File: src/smolagents/remote_executors.py [hunk 8]
```diff
@@ -249,6 +340,93 @@ def cleanup(self):
...
+            elif msg_type == "status" and msg_content["execution_state"] == "idle":
+                break
+
+        return CodeOutput(output=result, logs="".join(outputs), is_final_answer=is_final_answer)
+
+    except Exception as e:
+        logger.log_error(f"Code execution failed: {e}")
+        raise
+
+def _create_kernel_http(crate_kernel_endpoint: str, logger) -> str:
```
Comment (albertvillanova) [inline src/smolagents/remote_executors.py]: And the same for this one.

**Review by albertvillanova — COMMENTED** [review]
Thanks. I think there are some duplicated functions after having solved the conflicts with the main branch.

File: src/smolagents/remote_executors.py [hunk 9]
```diff
@@ -898,7 +990,7 @@ def delete(self):
                 try {
                   await micropip.install(packages);
                 } catch (e) {
-                  console.error(`Failed to load package ${pkg}: ${e.message}`);
+                  console.error(`Failed to load packages ${packages}: ${e.message}`);
```
Comment (albertvillanova) [inline src/smolagents/remote_executors.py]: Thanks for this fix! 🤗

File: src/smolagents/remote_executors.py [hunk 10]
```diff
@@ -52,6 +51,78 @@
     pass
 
 
+class JsonSerializer:
+    """Centralized JSON serialization/deserialization for remote executors."""
```
Comment (albertvillanova) [inline src/smolagents/remote_executors.py]: Much better implementation, without duplicated code. Thanks!

File: src/smolagents/remote_executors.py [hunk 11]
```diff
@@ -52,6 +51,78 @@
...
+
+    @staticmethod
+    def to_json_safe(obj):
+        """Convert Python objects to JSON-serializable format."""
+        if isinstance(obj, (str, int, float, bool, type(None))):
+            return obj
+        elif isinstance(obj, dict):
+            return {k: JsonSerializer.to_json_safe(v) for k, v in obj.items()}
```
Comment (albertvillanova) [inline src/smolagents/remote_executors.py]: I am wondering about the keys of the dict: they could be tuples, for example...
```python
data = {(1, 2): "value"}
```
I think we should call `JsonSerializer.to_json_safe` in the keys as well?

File: src/smolagents/remote_executors.py [hunk 12]
```diff
@@ -52,6 +51,78 @@
...
+        elif isinstance(obj, set):
+            return {"__type__": "set", "data": [JsonSerializer.to_json_safe(item) for item in obj]}
+        elif isinstance(obj, bytes):
+            return {"__type__": "bytes", "data": base64.b64encode(obj).decode()}
+        elif isinstance(obj, PIL.Image.Image):  # PIL Images
+            buffer = BytesIO()
+            obj.save(buffer, format="PNG")
+            return {"__type__": "PIL.Image", "data": base64.b64encode(buffer.getvalue()).decode()}
+        else:
+            raise TypeError(f"Object of type {type(obj).__name__} is not JSON serializable")
```
Comment (albertvillanova) [inline src/smolagents/remote_executors.py]: I see you removed from your previous implementation the support for `datetime` objects or `numpy` arrays. Why? More generally, which types should we support in this first version?

Additionally, as suggested in my first review, I think we could add some tests, like round-trip type fidelity validation. What do you think?

File: src/smolagents/remote_executors.py [hunk 13]
```diff
@@ -88,14 +159,29 @@ def send_tools(self, tools: dict[str, Tool]):
...
         if not variables:
             return
-        pickled_vars = base64.b64encode(pickle.dumps(variables)).decode()
+
+        # Convert variables to JSON-serializable format
+        json_safe_vars = JsonSerializer.to_json_safe(variables)
+        encoded_vars = base64.b64encode(json.dumps(json_safe_vars).encode()).decode()
+        ##
+        # do we even need base64 no that we have JSON serializing?! need to think about it
+        ##
```
Comment (albertvillanova) [inline src/smolagents/remote_executors.py]: I agree with you: I think base64 encoding is redundant here now:
- JSON is already text-safe and can be embedded in Python strings
- Adds around 33% size overhead
- Creates unnecessary encoding/decoding steps
```suggestion
        encoded_vars = json.dumps(json_safe_vars)
```

File: src/smolagents/remote_executors.py [hunk 14]
```diff
@@ -88,14 +159,29 @@ def send_tools(self, tools: dict[str, Tool]):
...
+import json, base64
+from io import BytesIO
+try:
+    import PIL.Image
+except ImportError:
+    PIL = None
+
+{JsonSerializer.get_from_json_safe_function_code()}
+
+vars_dict = json.loads(base64.b64decode('{encoded_vars}'))
```
Comment (albertvillanova) [inline src/smolagents/remote_executors.py]: I think we could remove base64 also here?

**Review by albertvillanova — CHANGES_REQUESTED** [review]
Thanks a lot: awesome work to eliminate pickle vulnerability!

The refactoring is excellent:
- Code duplication is eliminated: single `JsonSerializer` class with centralized logic
- Well-organized methods

I made some comments/suggestions below. Maybe, some of them could be addressed in subsequent PRs.

I think in this PR we need to fix:
- base64 is now unnecessary
- dict keys are not serialized
- maybe support numpy arrays, datetime objects,... and any other?

In future PRs we can:
- Add tests

What do you think?

File: src/smolagents/remote_executors.py (hunk 9 above)
Comment (TauroN3) [reply src/smolagents/remote_executors.py]: you welcome!

**Review by albertvillanova — APPROVED** [review]
Thank you for this awesome contribution! 🤗 

The pickle deserialization vulnerability is a critical security issue, and I really appreciate you taking the initiative to address it.

This PR establishes a secure, maintainable foundation for the serialization system. Any additional refinements or enhancements can be addressed incrementally in subsequent PRs as we gather user feedback and identify further improvements.

I've built upon your excellent foundation and added some changes to further improve the implementation.

Here's a comprehensive overview:
1. Security Fixes
    - Secure by default: Changed `allow_pickle=False` as the default in all agents and executors
    - Deprecation warnings: Added FutureWarning when pickle is used, guiding users toward safe alternatives
    - Fixed missing "pickle:" prefix: Updated main implementation to match documentation and generated code

    The vulnerability is now truly fixed with secure defaults. Users must explicitly opt-in to pickle usage.

2. API Simplification
    - Simplified API: Reduced from 2 confusing parameters → 1 clear parameter
      - Before: `allow_insecure_serializer=False`, `safe_serialization=True`
      - After: `allow_pickle=False`
    - More Pythonic naming: allow_pickle is self-documenting and follows Python conventions
    - Added "pickle:" prefix: For symmetry with "safe:" prefix in format detection

    50% simpler API that's intuitive and Pythonic.

3. Code Deduplication
    - Eliminated ~250 lines of duplication using code generation
    - Single source of truth: Used `inspect.getsource()` to generate remote code from actual implementation
 
    Much easier to maintain. Changes to serialization logic now propagate automatically.

4. Clean Architecture
    - Removed Tool inheritance: SafeSerializer is now a pure utility class

    Proper abstractions. Clear separation between agent tools and framework utilities.

5. Comprehensive Test Coverage
    - Security: Pickle blocked in safe mode
    - Roundtrip: All safe types work correctly
    - Optional dependencies: numpy, PIL support
    - Backward compatibility: Legacy pickle with explicit opt-in
    - Default behavior verification
    - Edge cases: Empty data, very large numbers, special floats (inf/nan), unicode, 1MB strings, deeply nested structures (100 levels), tuple/integer dict keys
    - Error handling: Invalid JSON, corrupted prefixes, missing type fields, unknown markers, invalid base64, unsupported types (functions, classes, modules)
    - Type coverage: All datetime types, Decimal precision, Path types, complex numbers, bytes/bytearray
    - etc.

    Comprehensive coverage proving the implementation works correctly in all scenarios.

---
Thank you again for the excellent work on identifying and addressing this security vulnerability! The improvements build on your solid foundation to create a production-ready, maintainable, and secure implementation.

**Comment by albertvillanova** [comment]
### Migration Path

For most users (recommended):
```python
# Just omit the parameter - secure by default!
agent = CodeAgent(model=model, tools=tools)
```

For legacy compatibility:
```python
# Explicit opt-in required
agent = CodeAgent(
    model=model, tools=tools,
    executor_type="docker"  # any of the remote executors: "blaxel", "e2b", "modal", "docker", "wasm"
    executor_kwargs={"allow_pickle": True}
)
# ⚠️  FutureWarning will be emitted
```
